### ISSUE TITLE
Optical Flow: allow uploading px4flow sensor firmware

### ISSUE THREAD

**Issue by rmackay9** [ArduPilot/MissionPlanner#743]
Allowing updating the px4flow sensor's firmware through the mission planner would be easier than users having to open qground control to do it.
https://pixhawk.org/modules/px4flow
We might have issues deciding which px4flow binary to use although i guess we could host the version that we think is best (i.e. that we've tested successfully) on firmware.diydrones.com

**Comment by meee1** [comment]
randy, is this on firmware.diydrones.com yet?

**Comment by rmackay9** [comment]
@meee1, ah right.  I didn't realise that was a requirement but makes perfect sense of course.

I'll talk to Tridge and we will put it somewhere.

**Comment by aerodolphin** [comment]
Any news about this?

**Comment by patrickpoirier51** [comment]
Michael, this issues need to be adressed because most  the new px4flow are shipped with a different firmware and we need a method to  roll back to the KLT version.

Its a little bit annoying to ask the end user to use QGC to flash the firmware and get back to Mission Planner to continue the config...

**Comment by meee1** [comment]
can add support, just need a copy of the older firmware. not sure where that is

**Comment by meee1** [comment]
the instructions are on the wiki
http://ardupilot.org/copter/docs/common-px4flow-overview.html

**Comment by patrickpoirier51** [comment]
Ok sorry for the noise

**Comment by patrickpoirier51** [comment]
Michael, I just tested the procedure and Mission planner and its not working well.
First of all the device is mapping twice:
![image](https://user-images.githubusercontent.com/12484190/35872171-27528eba-0b34-11e8-82e3-a3d38a11ce0f.png)
 
Just to make sure I deleted all ports and restarted Windows 10
So its mapping on COM3 and reload to COM4 (after 2 seconds)

When using the QGC to flash, we unplug-replug and the driver ''stops'' at COM3 and the firmware is uploaded correctly

When using MissionPlanner to flash , we select the comport (COM4) and then we  unplug - hit OK - Replug and then Mission planner stop responding...
I tried different methods but it allways end up with mission planner freezing or not finding the PX4FLOW.

**Comment by patrickpoirier51** [comment]
Michael,
I cant get the px4flow flashed with mission planner.
The driver is loaded and I can flash with QGC firmware if I specify the KLT firmware.
Here is my driver:
![image](https://user-images.githubusercontent.com/12484190/35894998-73600bbc-0b83-11e8-9c3d-6be8620ca5ea.png)

But according to the wiki I should see this:
![image](https://user-images.githubusercontent.com/12484190/35895031-a4ef099e-0b83-11e8-8dd1-d8919ffc0e3f.png)

Be I see this (No opticalflow definition on the serial device):
![image](https://user-images.githubusercontent.com/12484190/35895066-ca1ccd14-0b83-11e8-81e6-4be1a682f60c.png)

So when I try to flash, its looking for the device and cannot find it.

**Comment by rmackay9** [comment]
Using mission planner 1.3.52.10 I was able to upload the correct firmware.  I needed to enable installing unsigned drivers on my Win10 box before it worked.

**Comment by patrickpoirier51** [comment]
Seems that it is allways registering 2 instances  either with the painful driver signature disable .inf file or with the self installed px4 msi   http://pixhawk.org/static/px4driver.msi
I tried on both my px4flow sensors= same pattern, it is ''double mapping'' and Mission Planner can connect to the px4flow (adjust focus, etc), but it cant see it when trying to upload firmware.

![image](https://user-images.githubusercontent.com/12484190/35916208-a0e95bca-0bd7-11e8-94c7-40a91fb8ae91.png)

**Comment by meee1** [comment]
try using the Force PX4 Bootloader option

**Comment by meee1** [comment]
also you are showing devices that are not plugged in in your device manager.

**Comment by meee1** [comment]
also another method.
1. unplug px4flow
2. goto MP install FW
3. load custom firmware
4. press ok
 and plug board in.
5. done

**Comment by patrickpoirier51** [comment]
Michael,
 take note that the the px4flow maps as COM5  and then it reload (2 seconds)  it remaps as  COM13.
Yesterday I deleted all the USB drivers and it was mapping  COM3 and COM4 and I had the same pattern plug-map as com3 - reload - map as com4

Please note that this is in accordance with this:
https://pixhawk.org/users/px4flow_windows_driver

**My point is that QGC does the firmware upgrade no problem with the same environment.**

**Comment by meee1** [comment]
yep, the wiki probly needs an update to use the unplug replug method.

**Comment by patrickpoirier51** [comment]
Did your test with no success = it search without finding the px4flow...

Let me know if you need me to test other options and in the meantime I will suggest to use the signed PX4 driver 
http://pixhawk.org/static/px4driver.msi
and the QGC method for flashing firmware.
